File: week 5/edit_distance/edit_distance.cpp
#include <iostream>
#include <string>
#include <vector>

using std::string;
using std::vector;

int edit_distance(const string &str1, const string &str2) {
	int n = str1.size();
	int m = str2.size();
	vector<vector<int> > d(n + 1, vector<int>(m + 1, 0));
	for(int i = 0; i <= n; ++i) {
        d[i][0] = i;
    }
    for(int i = 0; i <= m; ++i) {
        d[0][i] = i;
    }
	int cost = 0;
	for(int i = 1; i <= n; ++i) {
		for(int j = 1; j <= m; ++j) {
			if(str1[i-1] == str2[j-1])
				cost = 0;
			else
				cost = 1;
			d[i][j] = std::min(std::min(d[i-1][j],d[i][j-1]) + 1, d[i-1][j-1] + cost);
		}
	}
	return d[n][m];
}

int main() {
  string str1;
  string str2;
  std::cin >> str1 >> str2;
  std::cout << edit_distance(str1, str2) << std::endl;
  return 0;
}

File: week%205/edit_distance/edit_distance.py
def edit_distance(s, t):
    n = len(s)
    m = len(t)
    # (n+1)rows-by-(m+1)cols matrix
    d = [[0 for x in range(m+1)] for y in range(n+1)] 
    for i in range(0, n+1):
        d[i][0] = i
    for i in range(0, m+1):
        d[0][i] = i
    cost = 0
    for i in range(1, n+1):
        for j in range(1, m+1):
            if(s[i-1] == t[j-1]):
                cost = 0
            else:
                cost = 1
            d[i][j] = min(min(d[i-1][j],d[i][j-1]) + 1, d[i-1][j-1] + cost)
    return d[n][m]

if __name__ == "__main__":
    print(edit_distance(raw_input(), raw_input()))